### ISSUE TITLE
Capitalization plus non-Vietnamese characters leads to inconsistent tokenization

### ISSUE THREAD

**Issue by Trey314159** [duydo/elasticsearch-analysis-vietnamese#26]
I see that the tokenizer splits on certain non-Vietnamese Latin characters, such as _å, ä, č, ç, ë, ğ, ï, ı, ñ, ö, ö, ø, š, ş, ü,_ and _ÿ_ (which is not a complete list, I expect, just the ones I tested). This is not ideal for text that contains foreign words, but it is also not uncommon for analyzers to split on foreign characters.

I also see that adjacent capitalized words are tokenized together, including names. This is also not always ideal in cases like _International Plant Names Index,_ but it is understandable.

However, these two interact unexpectedly; if the second letter of a capitalized word is a foreign character, it doesn't split. For example, the Spanish word _año_ (which means "year") is tokenized as _a + ñ + o,_ like this:
```
{
  "tokens" : [
    {
      "token" : "a",
      "start_offset" : 0,
      "end_offset" : 1,
      "type" : "phrase",
      "position" : 0
    },
    {
      "token" : "ñ",
      "start_offset" : 2,
      "end_offset" : 3,
      "type" : "residual",
      "position" : 1
    },
    {
      "token" : "o",
      "start_offset" : 4,
      "end_offset" : 5,
      "type" : "phrase",
      "position" : 2
    }
  ]
}
```
However, the capitalized form, _Año_ is tokenized as _añ + o,_ like this:
```
{
  "tokens" : [
    {
      "token" : "añ",
      "start_offset" : 0,
      "end_offset" : 2,
      "type" : "allcaps",
      "position" : 0
    },
    {
      "token" : "o",
      "start_offset" : 3,
      "end_offset" : 4,
      "type" : "phrase",
      "position" : 1
    }
  ]
}
```

It seems that capitalization should not change the tokenization of a single word.

I'm also seeing similar behavior with other splitting characters: _r=22_ → _r + 22,_ but _R=22_ → _r= + 22._ Similarly, _x&y_ vs _X&Y,_ _x*y_ vs _X*Y,_ _x+y_ vs _X+Y,_ and others.

**Comment by duydo** [comment]
@Trey314159 Thank you very much for this report, the tokenizer should not analyze the foreign words.
It seems that we have a lot of issues with the Vietnamese tokenizer library, I used a very old version for the plugin (released in 2009). I tried to replace this old one with the latest one here https://github.com/phuonglh/vn.vitk but the performance was very poor, because of Spark.
In order to use the new one with Elasticsearch, I think we should reimplement the tokenizer without Spark. What do you think?

**Comment by Trey314159** [comment]
Oddly, I'm getting messages from @duydo in email, but not seeing them here. Not sure where/how to respond properly.

**Comment by duydo** [comment]
Hi @Trey314159,

I've just fixed the offset issue #25 and implemented your suggestions for the tokenizer. Now the tokenizer does not split non-Vietnamese characters, does not tokenize the capitalized words together, split words/character correctly.

For the text `sách .. sách ; 東京都`:
```
{
  "tokens" : [
    {
      "token" : "sách",
      "start_offset" : 0,
      "end_offset" : 4,
      "type" : "<PHRASE>",
      "position" : 0
    },
    {
      "token" : "sách",
      "start_offset" : 8,
      "end_offset" : 12,
      "type" : "<PHRASE>",
      "position" : 1
    },
    {
      "token" : "東京都",
      "start_offset" : 15,
      "end_offset" : 18,
      "type" : "<FOREIGN>",
      "position" : 2
    }
  ]
}
```

For the text `año vs Año, r=22 vs R=22, x&y vs X&Y, x*y vs X*Y, x+y vs X+Y`:
```
{
  "tokens" : [
    {
      "token" : "año",
      "start_offset" : 0,
      "end_offset" : 3,
      "type" : "<PHRASE>",
      "position" : 0
    },
    {
      "token" : "vs",
      "start_offset" : 4,
      "end_offset" : 6,
      "type" : "<PHRASE>",
      "position" : 1
    },
    {
      "token" : "año",
      "start_offset" : 7,
      "end_offset" : 10,
      "type" : "<PHRASE>",
      "position" : 2
    },
    {
      "token" : "r",
      "start_offset" : 12,
      "end_offset" : 13,
      "type" : "<PHRASE>",
      "position" : 3
    },
    {
      "token" : "22",
      "start_offset" : 14,
      "end_offset" : 16,
      "type" : "<NUMBER>",
      "position" : 4
    },
    {
      "token" : "vs",
      "start_offset" : 17,
      "end_offset" : 19,
      "type" : "<PHRASE>",
      "position" : 5
    },
    {
      "token" : "r",
      "start_offset" : 20,
      "end_offset" : 21,
      "type" : "<PHRASE>",
      "position" : 6
    },
    {
      "token" : "22",
      "start_offset" : 22,
      "end_offset" : 24,
      "type" : "<NUMBER>",
      "position" : 7
    },
    {
      "token" : "x",
      "start_offset" : 26,
      "end_offset" : 27,
      "type" : "<PHRASE>",
      "position" : 8
    },
    {
      "token" : "y",
      "start_offset" : 28,
      "end_offset" : 29,
      "type" : "<PHRASE>",
      "position" : 9
    },
    {
      "token" : "vs",
      "start_offset" : 30,
      "end_offset" : 32,
      "type" : "<PHRASE>",
      "position" : 10
    },
    {
      "token" : "x",
      "start_offset" : 33,
      "end_offset" : 34,
      "type" : "<PHRASE>",
      "position" : 11
    },
    {
      "token" : "y",
      "start_offset" : 35,
      "end_offset" : 36,
      "type" : "<PHRASE>",
      "position" : 12
    },
    {
      "token" : "x",
      "start_offset" : 38,
      "end_offset" : 39,
      "type" : "<PHRASE>",
      "position" : 13
    },
    {
      "token" : "y",
      "start_offset" : 40,
      "end_offset" : 41,
      "type" : "<PHRASE>",
      "position" : 14
    },
    {
      "token" : "vs",
      "start_offset" : 42,
      "end_offset" : 44,
      "type" : "<PHRASE>",
      "position" : 15
    },
    {
      "token" : "x",
      "start_offset" : 45,
      "end_offset" : 46,
      "type" : "<PHRASE>",
      "position" : 16
    },
    {
      "token" : "y",
      "start_offset" : 47,
      "end_offset" : 48,
      "type" : "<PHRASE>",
      "position" : 17
    },
    {
      "token" : "x",
      "start_offset" : 50,
      "end_offset" : 51,
      "type" : "<PHRASE>",
      "position" : 18
    },
    {
      "token" : "y",
      "start_offset" : 52,
      "end_offset" : 53,
      "type" : "<PHRASE>",
      "position" : 19
    },
    {
      "token" : "vs",
      "start_offset" : 54,
      "end_offset" : 56,
      "type" : "<PHRASE>",
      "position" : 20
    },
    {
      "token" : "x",
      "start_offset" : 57,
      "end_offset" : 58,
      "type" : "<PHRASE>",
      "position" : 21
    },
    {
      "token" : "y",
      "start_offset" : 59,
      "end_offset" : 60,
      "type" : "<PHRASE>",
      "position" : 22
    }
  ]
}
```
I attached here the build for your ES v5.3.2 
[elasticsearch-analysis-vietnamese-5.3.2.zip](https://github.com/duydo/elasticsearch-analysis-vietnamese/files/1177209/elasticsearch-analysis-vietnamese-5.3.2.zip), it would be great if you can test it again.

Thank you very much for your contribution.
- Duy

**Comment by Trey314159** [comment]
@duydo, thanks for the update! I've downloaded the zip and I'll try to give it a thorough testing next week and report back here.

**Comment by duydo** [comment]
@Trey314159 I just found a bug when doing the query, please don't test that version util it's fixed.

**Comment by Trey314159** [comment]
@duydo, let me know when there's a new version.

**Comment by duydo** [comment]
Hi @Trey314159,
After investigating I found that the original tokenizer is not thread-safe, it's a root cause of the NullPointerException as you reported. I've just re-implemented a new thread-safe tokenizer, can you please give it a try? Here is the new version for ES v5.3.2 [elasticsearch-analysis-vietnamese-5.3.2.zip](https://github.com/duydo/elasticsearch-analysis-vietnamese/files/1183610/elasticsearch-analysis-vietnamese-5.3.2.zip).

Again, thank you very much for raising this critical issue :-)

**Comment by Trey314159** [comment]
Cool! I've downloaded it and I'll give it a try next week.

**Comment by duydo** [comment]
I close this issue, it will be fixed in new tokenizer #37
